core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -235,41 +242,71 @@ object AbstractFetcherThread {
 
 }
 
+object FetcherMetrics {
+  val ConsumerLag = "ConsumerLag"
+  val RequestsPerSec = "RequestsPerSec"
+  val BytesPerSec = "BytesPerSec"
+}
+
 class FetcherLagMetrics(metricId: ClientIdTopicPartition) extends KafkaMetricsGroup {
+
   private[this] val lagVal = new AtomicLong(-1L)
-  newGauge("ConsumerLag",
+  private[this] val tags = Map(
+    "clientId" -> metricId.clientId,
+    "topic" -> metricId.topic,
+    "partition" -> metricId.partitionId.toString)
+
+  newGauge(FetcherMetrics.ConsumerLag,
     new Gauge[Long] {
       def value = lagVal.get
     },
-    Map("clientId" -> metricId.clientId,
-      "topic" -> metricId.topic,
-      "partition" -> metricId.partitionId.toString)
+    tags
   )
 
   def lag_=(newLag: Long) {
     lagVal.set(newLag)
   }
 
   def lag = lagVal.get
+
+  def unregister() {
+    removeMetric(FetcherMetrics.ConsumerLag, tags)
+  }
 }
 
 class FetcherLagStats(metricId: ClientIdAndBroker) {
   private val valueFactory = (k: ClientIdTopicPartition) => new FetcherLagMetrics(k)
   val stats = new Pool[ClientIdTopicPartition, FetcherLagMetrics](Some(valueFactory))
 
-  def getFetcherLagStats(topic: String, partitionId: Int): FetcherLagMetrics = {
+  def getAndMaybePut(topic: String, partitionId: Int): FetcherLagMetrics = {
     stats.getAndMaybePut(new ClientIdTopicPartition(metricId.clientId, topic, partitionId))
   }
+
+  def unregister(topic: String, partitionId: Int) {
+    val lagMetrics = stats.remove(new ClientIdTopicPartition(metricId.clientId, topic, partitionId))
+    if (lagMetrics != null) lagMetrics.unregister()
+  }
+
+  def unregister() {
+    stats.values.foreach(_.unregister())
+    stats.clear()
```

Comment:
I wonder if we should write this defensively, something like:

```scala
stats.keys.toBuffer.foreach { key =>
  unregister(key.topic, key.partitionId)
}
```

This would mean that the method could be called from multiple threads and we would not potentially leak metrics. It's a bit overkill since this is only called from shutdown, but seems harmless. Otherwise, we should probably add a comment saying that `unregister` is not safe to be called from multiple threads concurrently.